### PR TITLE
WIP refactor(saber): add node-notifier

### LINKED ISSUE #509: Show a notifier when it's required to restart server
#### What problem does this feature solve?

https://github.com/saberland/saber/blob/83c18a706637fedccdbe0ad34566218dc5780d94/packages/saber/lib/plugins/extend-node-api.js#L89

People often modify `saber-config.js`, `saber-node.js` and forget to restart server.

#### How should this be implemented in your opinion?

Maybe use https://www.npmjs.com/package/node-notifier, and expose it in `saber-utils` package

### PR DESCRIPTION
**Pull request by tyankatsu0105**
-  [x] Add `node-notifier` with `yarn workspace`
- [ ] ~~Write test~~ Not required
- [x] Add icon for `packages/assets`
-  [x] Refactor exist code

**Summary**

close #509 

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [x] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI/CSS related code, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] Briefly describing what this PR does in the title in [Angular Commit Message Conventions](https://git.io/fNGDG), e.g. `fix: rebuild when a page is added`

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

**Comment by tyankatsu0105** [comment]
@egoist 
I don't know how to test :_(

File: packages/saber/lib/plugins/extend-node-api.js [hunk 1]
```diff
@@ -86,7 +88,7 @@ exports.apply = api => {
           )
           // Because you might also update webpack config in saber-node.js
           // Which we can't (?) automatically reload
-          log.warn(`You probably need to restart the server.`)
+          notifier.notify('You probably need to restart the server.')
```
Comment (egoist) [inline packages/saber/lib/plugins/extend-node-api.js]: we should still output warning in console

Comment (tyankatsu0105) [reply packages/saber/lib/plugins/extend-node-api.js]: Oops. sorry.

**Comment by tyankatsu0105** [comment]
@MarchWorks 
Thanks for the reply.
But I don't understand what you say...
Why do I add some comments that file?

**Comment by MarchWorks** [comment]
@tyankatsu0105 sorry if I am mistaken but you said `I don't know how to test :_(` so if you follow these steps
```
# Link the `saber` binary
cd packages/saber
yarn link

# If you wanna run the website locally
cd website
yarn
yarn dev # which uses the global `saber` command
```
then you add a comment here https://github.com/saberland/saber/blob/master/website/saber-node.js you will trigger the warning and the notification.

File: packages/saber/lib/plugins/extend-node-api.js (hunk 1 above)
Comment (MarchWorks) [reply packages/saber/lib/plugins/extend-node-api.js]: We should set `Saber` as the title

**Comment by egoist** [comment]
We can also display the Saber logo in the notifier https://github.com/saberland/art/blob/master/icon/icon-alt.png

**Comment by egoist** [comment]
> extend-node-api.js can't import icon from saberland/art.
> Can I put icon for plugins directory or something?

You can put it inside `saber/assets` directory, also include that folder in the `files` array in `package.json`

**Comment by egoist** [comment]
We should make the message more detailed, like `saber-node.js was changed, you need to restart the server`

**Comment by tyankatsu0105** [comment]
OK! 
Thanks, @egoist .
I'll try doing that.

**Comment by tyankatsu0105** [comment]
@egoist 
I'm finished!
Would you mind reviewing this PR?

**Comment by krmax44** [comment]
What about including a button in the notification that allows restarting directly?

**Comment by tyankatsu0105** [comment]
>button in the notification that allows restarting directly

That’s interesting.
I'll try doing that.

**Comment by egoist** [comment]
1. Just checked the install size of node-notifier: 5.5MB, now I'm not sure if it is worth adding just for two warnings 🤔

2. Not quite happy about the result of node-notifier, can we get rid of that default icon?

image

3. Showing a notifier on macOS Catalina will request user permission, not sure if it's a good DX, what about showing the warning elsewhere? Like in the browser 🤔

**Comment by egoist** [comment]
> @krmax44: What about including a button in the notification that allows restarting directly?

That would require Saber to be run in a child process, Next.js used to auto-restart when config file is changed but it seems [they removed this feature](https://github.com/zeit/next.js/pull/6450/files#diff-2a8b3f9d55a8606efc8dedba1f423fc3L125-L138).

**Comment by tyankatsu0105** [comment]
Sorry, I don't know how to restart server :_(

**Comment by tyankatsu0105** [comment]
Oh. I just checked [this comment](https://github.com/saberland/saber/pull/514#issuecomment-541424191). 
I think this feature is not needed for saber.

>Like in the browser
Um, could you give me an example?

**Comment by tyankatsu0105** [comment]
@egoist 
Thank you for your several opinions.
If you think unnecessary this feature, please close this PR. And then, let's open a new issue.

**Comment by MarchWorks** [comment]
@tyankatsu0105 check this PR https://github.com/saberland/saber/pull/503 as an example of sending a notification to the browser. The difference is that you will have to send your own data from the development server (maybe it's possible to do it with webpack-dev-middleware but I am not sure how)

**Comment by tyankatsu0105** [comment]
I close this PR.
Because this PR's purpose is incorrect for the issue.
re https://github.com/saberland/saber/pull/514#issuecomment-541421199

Maybe `node-notifier` doesn't meet the needs.

**Comment by tyankatsu0105** [comment]
Sorry that I couldn't be your help.  :_(

**Pull request closed without merging**